File: temp/source/pipeline/sources/ainews.ts
// AI News source - extracts sub-stories from the news.smol.ai RSS feed.
//
// AI News is a daily aggregator of X/Twitter, Reddit, and Discord AI content.
// Each issue has two useful sections:
//   - <h1>AI Twitter Recap</h1>   - one editorial top-story per issue
//   - <h1>AI Reddit Recap</h1>    - multiple reddit posts, one per <a reddit.com/…/comments/…>
//
// We turn each issue into multiple FetchedArticle entries:
//   - One twitter-recap sub-story (title = first <strong>, primary URL = best
//     non-tweet link, content = first few bullets as summary)
//   - N reddit-recap sub-stories (one per reddit post anchor)
//
// Primary-URL priority: github > huggingface > arxiv > blog > reddit > tweet.
//
// This file is deliberately self-contained and does NOT reuse email.ts logic.

import type { FetchedArticle } from "@shared/types";
import { log } from "../utils";
import { fetchWithTimeout, isWithinWindow } from "./utils";

const FEED_URL = "https://news.smol.ai/rss.xml";
const MAX_CONTENT_LENGTH = 1400;
const MIN_REDDIT_BODY_LENGTH = 120;
const MIN_TWITTER_TITLE_LENGTH = 10;

type PrimaryKind =
  | "tweet"
  | "reddit"
  | "github"
  | "huggingface"
  | "arxiv"
  | "blog"
  | "other";

// ---------- feed parsing (no deps) ----------

const HTML_ENTITIES: Record<string, string> = {
  "&lt;": "<",
  "&gt;": ">",
  "&quot;": '"',
  "&apos;": "'",
  "&#x27;": "'",
  "&#x26;": "&",
  "&nbsp;": " ",
  "&amp;": "&",
};

function decodeHtml(s: string): string {
  return s.replace(
    /&(?:lt|gt|quot|apos|#x27|#x26|nbsp|amp);/g,
    (m) => HTML_ENTITIES[m],
  );
}

function pickTag(block: string, tag: string): string {
  const match = block.match(new RegExp(`<${tag}>([\\s\\S]*?)</${tag}>`));
  if (!match) return "";
  const cdata = match[1].match(/^<!\[CDATA\[([\s\S]*)\]\]>$/);
  return cdata ? cdata[1] : match[1];
}

interface RssItem {
  title: string;
  link: string;
  contentHtml: string;
  pubDate: string;
}

function parseFeed(xml: string): RssItem[] {
  return xml
    .split("<item>")
    .slice(1)
    .map((chunk) => {
      const body = chunk.split("</item>")[0];
      return {
        title: pickTag(body, "title"),
        link: pickTag(body, "link"),
        contentHtml: decodeHtml(pickTag(body, "content:encoded")),
        pubDate: pickTag(body, "pubDate"),
      };
    });
}

// ---------- HTML helpers ----------

function stripTags(html: string): string {
  return html
    .replace(/<[^>]+>/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function firstBold(html: string): string {
  const match = html.match(/<strong[^>]*>([\s\S]*?)<\/strong>/);
  return match ? stripTags(match[1]) : "";
}

interface Link {
  url: string;
  host: string;
  text: string;
}

function hostOf(url: string): string {
  try {
    return new URL(url).hostname.replace(/^www\./, "");
  } catch {
    return "";
  }
}

function collectLinks(html: string): Link[] {
  const matches = html.matchAll(/<a[^>]+href="([^"]+)"[^>]*>([\s\S]*?)<\/a>/g);
  return [...matches].map((m) => ({
    url: m[1],
    host: hostOf(m[1]),
    text: stripTags(m[2]),
  }));
}

/**
 * Extract the HTML region between a given <h1> and the next <h1>.
 * Returns an empty string when the heading is not present.
 */
function sliceSection(html: string, headingRx: RegExp): string {
  const start = html.search(headingRx);
  if (start === -1) return "";
  const afterStart = html.slice(start).replace(headingRx, "");
  const nextH1 = afterStart.search(/<h1[^>]*>/i);
  return nextH1 === -1 ? afterStart : afterStart.slice(0, nextH1);
}

// ---------- primary-URL priority ----------

const BLOG_HOST_RX =
  /\b(openai|anthropic|deepmind|google|meta|microsoft|nvidia|mistral|unsloth|databricks|cohere|stability|perplexity|vercel|modal|together|replicate|groq)\b/;

function classifyHost(host: string): PrimaryKind | null {
  if (host === "github.com" || host.endsWith(".github.io")) return "github";
  if (host === "huggingface.co" || host === "hf.co") return "huggingface";
  if (host === "arxiv.org" || host === "ar5iv.labs.arxiv.org") return "arxiv";
  if (
    BLOG_HOST_RX.test(host) ||
    host.endsWith(".ai") ||
    host.endsWith(".dev") ||
    host.includes("substack.com")
  ) {
    return "blog";
  }
  if (host === "reddit.com" || host.endsWith(".reddit.com")) return "reddit";
  if (host === "x.com" || host === "twitter.com") return "tweet";
  return null;
}

// Priority order applied when picking a primary link for a sub-story.
const PRIMARY_PRIORITY: PrimaryKind[] = [
  "github",
  "huggingface",
  "arxiv",
  "blog",
  "reddit",
  "tweet",
];

function isJunk(link: Link): boolean {
  if (!link.url) return true;
  if (
    link.host === "news.smol.ai" ||
    link.host === "latent.space" ||
    link.host === "support.substack.com" ||
    link.host === "i.redd.it" ||
    link.host === "preview.redd.it"
  ) {
    return true;
  }
  if (link.url.includes("twitter.com/i/lists/")) return true;
  if (link.url.includes("x.com/i/")) return true;
  return false;
}

function pickPrimary(links: Link[]): { url: string; kind: PrimaryKind } {
  const clean = links.filter((l) => !isJunk(l));
  if (clean.length === 0) return { url: "", kind: "other" };

  for (const kind of PRIMARY_PRIORITY) {
    const hit = clean.find((l) => classifyHost(l.host) === kind);
    if (hit) return { url: hit.url, kind };
  }
  return { url: clean[0].url, kind: "other" };
}

// ---------- Twitter Recap (one sub-story per issue) ----------

interface SubStory {
  title: string;
  url: string;
  content: string;
}

function extractTwitterRecap(html: string): SubStory | null {
  const region = sliceSection(html, /<h1[^>]*>\s*AI Twitter Recap\s*<\/h1>/i);
  if (!region) return null;

  const title = firstBold(region);
  if (!title || title.length < MIN_TWITTER_TITLE_LENGTH) return null;

  const { url } = pickPrimary(collectLinks(region));
  if (!url) return null;

  // Use the first few bullets as the summary instead of the full recap.
  const firstBullets = [...region.matchAll(/<li>([\s\S]*?)<\/li>/g)]
    .slice(0, 3)
    .map((m) => stripTags(m[1]))
    .join(" ");
  const summary =
    firstBullets || stripTags(region).slice(0, MAX_CONTENT_LENGTH);

  return { title, url, content: summary.slice(0, MAX_CONTENT_LENGTH) };
}

// ---------- Reddit Recap (one sub-story per post anchor) ----------

const REDDIT_ANCHOR_RX =
  /<a[^>]+href="(https?:\/\/(?:www\.)?reddit\.com\/r\/[^"]+\/comments\/[^"]+)"[^>]*>([\s\S]*?)<\/a>/g;

interface RedditAnchor {
  url: string;
  title: string;
  start: number;
  end: number;
}

function findRedditAnchors(region: string): RedditAnchor[] {
  const anchors: RedditAnchor[] = [];
  for (const match of region.matchAll(REDDIT_ANCHOR_RX)) {
    anchors.push({
      url: match[1],
      title: stripTags(match[2]),
      start: match.index ?? 0,
      end: (match.index ?? 0) + match[0].length,
    });
  }
  return anchors;
}

function extractRedditRecap(html: string): SubStory[] {
  const region = sliceSection(html, /<h1[^>]*>\s*AI Reddit Recap\s*<\/h1>/i);
  if (!region) return [];

  const anchors = findRedditAnchors(region);
  const stories: SubStory[] = [];

  for (let i = 0; i < anchors.length; i++) {
    const anchor = anchors[i];
    const nextAnchorStart =
      i + 1 < anchors.length ? anchors[i + 1].start : region.length;

    // Stop body at next anchor OR next h2/h3, whichever comes first.
    const afterAnchor = region.slice(anchor.end);
    const headingRel = afterAnchor.search(/<h[123][^>]*>/i);
    const headingAbs =
      headingRel === -1 ? region.length : anchor.end + headingRel;

    const bodyEnd = Math.min(nextAnchorStart, headingAbs);
    const body = region.slice(anchor.start, bodyEnd);
    const content = stripTags(body);
    if (content.length < MIN_REDDIT_BODY_LENGTH) continue;

    const { url } = pickPrimary(collectLinks(body));
    stories.push({
      title: anchor.title || "(untitled)",
      url: url || anchor.url,
      content: content.slice(0, MAX_CONTENT_LENGTH),
    });
  }
  return stories;
}

// ---------- fetch ----------

async function fetchFeedXml(): Promise<string | null> {
  try {
    const res = await fetchWithTimeout(FEED_URL);
    if (!res.ok) throw new Error(`Status ${res.status}`);
    return await res.text();
  } catch (err) {
    log(`  AI News fetch FAILED: ${err}`);
    return null;
  }
}

function toArticle(story: SubStory, pubDate: string): FetchedArticle {
  return {
    title: story.title,
    url: story.url,
    content: story.content,
    publishedDate: pubDate,
    source: "AI News",
    sourceType: "aiNews",
  };
}

export async function fetchAiNews(): Promise<FetchedArticle[]> {
  log("Fetching AI News feed...");

  const xml = await fetchFeedXml();
  if (!xml) return [];

  const items = parseFeed(xml).filter((it) => isWithinWindow(it.pubDate));
  log(`  AI News: ${items.length} issues within window`);

  const articles: FetchedArticle[] = [];
  for (const item of items) {
    // Skip low-signal days the editors tag as "not much happened".
    if (/^not much happened/i.test(item.title)) continue;

    const twitter = extractTwitterRecap(item.contentHtml);
    if (twitter) articles.push(toArticle(twitter, item.pubDate));

    for (const reddit of extractRedditRecap(item.contentHtml)) {
      articles.push(toArticle(reddit, item.pubDate));
    }
  }

  // Dedup by URL (same repo/post referenced in multiple issues).
  const byUrl = new Map<string, FetchedArticle>();
  for (const article of articles) {
    if (!byUrl.has(article.url)) byUrl.set(article.url, article);
  }
  const deduped = [...byUrl.values()];

  log(`AI News: ${deduped.length} articles extracted`);
  return deduped;
}

File: backend/pipeline/sources/ainews.py
"""AI News source - extracts sub-stories from the news.smol.ai RSS feed."""

from __future__ import annotations

import re
from html import unescape

from pipeline.sources.utils import fetch_with_timeout, is_within_window
from pipeline.utils import log

FEED_URL = "https://news.smol.ai/rss.xml"
MAX_CONTENT_LENGTH = 1400
MIN_REDDIT_BODY_LENGTH = 120
MIN_TWITTER_TITLE_LENGTH = 10


# --- minimal RSS parsing (no feedparser dep for this source) ---


def _pick_tag(block: str, tag: str) -> str:
    match = re.search(rf"<{tag}>([\s\S]*?)</{tag}>", block)
    if not match:
        return ""
    val = match.group(1)
    cdata = re.match(r"^<!\[CDATA\[([\s\S]*)\]\]>$", val)
    return cdata.group(1) if cdata else val


def _parse_feed(xml: str) -> list[dict]:
    items = []
    for chunk in xml.split("<item>")[1:]:
        body = chunk.split("</item>")[0]
        items.append(
            {
                "title": _pick_tag(body, "title"),
                "link": _pick_tag(body, "link"),
                "content_html": unescape(_pick_tag(body, "content:encoded")),
                "pub_date": _pick_tag(body, "pubDate"),
            }
        )
    return items


# --- HTML helpers ---


def _strip_tags(html: str) -> str:
    return re.sub(r"\s+", " ", re.sub(r"<[^>]+>", " ", html)).strip()


def _first_bold(html: str) -> str:
    m = re.search(r"<strong[^>]*>([\s\S]*?)</strong>", html)
    return _strip_tags(m.group(1)) if m else ""


def _host_of(url: str) -> str:
    try:
        from urllib.parse import urlparse

        return urlparse(url).hostname.removeprefix("www.")
    except Exception:
        return ""


def _collect_links(html: str) -> list[dict]:
    return [
        {
            "url": m.group(1),
            "host": _host_of(m.group(1)),
            "text": _strip_tags(m.group(2)),
        }
        for m in re.finditer(r'<a[^>]+href="([^"]+)"[^>]*>([\s\S]*?)</a>', html)
    ]


def _slice_section(html: str, heading_rx: re.Pattern) -> str:
    start = heading_rx.search(html)
    if not start:
        return ""
    after = html[start.end() :]
    next_h1 = re.search(r"<h1[^>]*>", after, re.IGNORECASE)
    return after[: next_h1.start()] if next_h1 else after


BLOG_HOST_RX = re.compile(
    r"\b(openai|anthropic|deepmind|google|meta|microsoft|nvidia|mistral|unsloth|"
    r"databricks|cohere|stability|perplexity|vercel|modal|together|replicate|groq)\b"
)

PRIMARY_PRIORITY = ["github", "huggingface", "arxiv", "blog", "reddit", "tweet"]

_JUNK_HOSTS = {
    "news.smol.ai",
    "latent.space",
    "support.substack.com",
    "i.redd.it",
    "preview.redd.it",
}


def _classify_host(host: str) -> str | None:
    if host in ("github.com",) or host.endswith(".github.io"):
        return "github"
    if host in ("huggingface.co", "hf.co"):
        return "huggingface"
    if host in ("arxiv.org", "ar5iv.labs.arxiv.org"):
        return "arxiv"
    if (
        BLOG_HOST_RX.search(host)
        or host.endswith(".ai")
        or host.endswith(".dev")
        or "substack.com" in host
    ):
        return "blog"
    if host in ("reddit.com",) or host.endswith(".reddit.com"):
        return "reddit"
    if host in ("x.com", "twitter.com"):
        return "tweet"
    return None


def _is_junk(link: dict) -> bool:
    if not link["url"]:
        return True
    if link["host"] in _JUNK_HOSTS:
        return True
    if "twitter.com/i/lists/" in link["url"] or "x.com/i/" in link["url"]:
        return True
    return False


def _pick_primary(links: list[dict]) -> dict:
    clean = [lnk for lnk in links if not _is_junk(lnk)]
    if not clean:
        return {"url": "", "kind": "other"}
    for kind in PRIMARY_PRIORITY:
        hit = next((lnk for lnk in clean if _classify_host(lnk["host"]) == kind), None)
        if hit:
            return {"url": hit["url"], "kind": kind}
    return {"url": clean[0]["url"], "kind": "other"}


# --- Twitter Recap ---


def _extract_twitter_recap(html: str) -> dict | None:
    region = _slice_section(
        html, re.compile(r"<h1[^>]*>\s*AI Twitter Recap\s*</h1>", re.IGNORECASE)
    )
    if not region:
        return None
    title = _first_bold(region)
    if not title or len(title) < MIN_TWITTER_TITLE_LENGTH:
        return None
    primary = _pick_primary(_collect_links(region))
    if not primary["url"]:
        return None
    bullets = [
        _strip_tags(m.group(1)) for m in re.finditer(r"<li>([\s\S]*?)</li>", region)
    ][:3]
    content = " ".join(bullets) or _strip_tags(region)
    return {
        "title": title,
        "url": primary["url"],
        "content": content[:MAX_CONTENT_LENGTH],
    }


# --- Reddit Recap ---

REDDIT_ANCHOR_RX = re.compile(
    r'<a[^>]+href="(https?://(?:www\.)?reddit\.com/r/[^"]+/comments/[^"]+)"[^>]*>([\s\S]*?)</a>'
)


def _extract_reddit_recap(html: str) -> list[dict]:
    region = _slice_section(
        html, re.compile(r"<h1[^>]*>\s*AI Reddit Recap\s*</h1>", re.IGNORECASE)
    )
    if not region:
        return []

    anchors = [
        {
            "url": m.group(1),
            "title": _strip_tags(m.group(2)),
            "start": m.start(),
            "end": m.end(),
        }
        for m in REDDIT_ANCHOR_RX.finditer(region)
    ]

    stories = []
    for i, anchor in enumerate(anchors):
        next_start = anchors[i + 1]["start"] if i + 1 < len(anchors) else len(region)
        after = region[anchor["end"] :]
        heading_rel = re.search(r"<h[123][^>]*>", after, re.IGNORECASE)
        heading_abs = (
            anchor["end"] + heading_rel.start() if heading_rel else len(region)
        )
        body_end = min(next_start, heading_abs)
        body = region[anchor["start"] : body_end]
        content = _strip_tags(body)
        if len(content) < MIN_REDDIT_BODY_LENGTH:
            continue
        primary = _pick_primary(_collect_links(body))
        stories.append(
            {
                "title": anchor["title"] or "(untitled)",
                "url": primary["url"] or anchor["url"],
                "content": content[:MAX_CONTENT_LENGTH],
            }
        )
    return stories


# --- fetch ---


def fetch_ai_news() -> list[dict]:
    log("Fetching AI News feed...")
    try:
        resp = fetch_with_timeout(FEED_URL)
        if not resp.is_success:
            raise ValueError(f"Status {resp.status_code}")
        xml = resp.text
    except Exception as e:
        log(f"  AI News fetch FAILED: {e}")
        return []

    items = [it for it in _parse_feed(xml) if is_within_window(it["pub_date"])]
    log(f"  AI News: {len(items)} issues within window")

    articles: list[dict] = []
    for item in items:
        if re.match(r"^not much happened", item["title"], re.IGNORECASE):
            continue
        twitter = _extract_twitter_recap(item["content_html"])
        if twitter:
            articles.append(
                {
                    **twitter,
                    "published_date": item["pub_date"],
                    "source": "AI News",
                    "source_type": "aiNews",
                }
            )
        for reddit in _extract_reddit_recap(item["content_html"]):
            articles.append(
                {
                    **reddit,
                    "published_date": item["pub_date"],
                    "source": "AI News",
                    "source_type": "aiNews",
                }
            )

    by_url: dict[str, dict] = {}
    for a in articles:
        if a["url"] not in by_url:
            by_url[a["url"]] = a

    log(f"AI News: {len(by_url)} articles extracted")
    return list(by_url.values())
